libraries/core_libs/network/src/tarcap/packets_handlers/pbft_sync_packet_handler.cpp

```diff
@@ -152,6 +154,24 @@ void PbftSyncPacketHandler::process(const PacketData &packet_data, const std::sh
     return;
   }
 
+  // This is special case when queue is empty and we can not say for sure that all votes that are part of this block
+  // have been verified before
+  if (pbft_mgr_->periodDataQueueEmpty()) {
+    for (const auto &v : period_data.previous_block_cert_votes) {
+      vote_mgr_->addRewardVote(v);
+    }
+    if (!vote_mgr_->checkRewardVotes(period_data.pbft_blk)) {
+      LOG(log_er_) << "Invalid reward votes in block " << period_data.pbft_blk->getBlockHash() << " from peer "
+                   << packet_data.from_node_id_.abridged() << " received, stop syncing.";
+      handleMaliciousSyncPeer(packet_data.from_node_id_);
+      return;
+    }
+    // And now we need to replace it with verified votes
+    if (auto votes = pbft_mgr_->getRewardVotesInBlock(period_data.pbft_blk->getRewardVotes()); votes.size()) {
```

Comment:
I know this is not your code, on the other hand you approved it... from this codeblock it is pretty obvious that the design for reward votes is bad. We are adding reward votes to some container through `addRewardVote` just because this function actually verifies the vote, then we check if those votes were added into the container through `checkRewardVotes` and finally we get them from it through `getRewardVotesInBlock`. Most of this processing is totally useless...

What we actually need is just verify the reward votes and thats it. Few comments how to achieve it:

- separate addRewardVote and verifyRewardVote
- put those few checks that are inside addRewardVote to verifyRewardVote
- I would also rather call verifyRewardVote explicitly before calling addRewardVote (optional)
- rename checkRewardVotes to something more self explanatory - it actually just checks if votes are inside the container
- in this codeblock you would then call just verifyRewardVote

We can either create separate issue for it, or you can fix it in this PR...

Reply:
I do not agree with everything, e.g. :

- addRewardVote is checking if reward is already in cache and do not verified it multiple times, so current design for that is ok, just name could be addAndVerify?
- not sure what you have against `checkRewardVotes` it takes pbft block as arg, so I understand exactly what is doing it is checking reward votes (that are **always** inside the container) and check if it matches required ones by block

Reply:
This is nice straightforward design that is easy to read and also adjust later, we also use this pattern in other packet handlers (not having it in 1 function called addRewardVote as now):

for (vote : votes) {
 if (rewards_votes.isKnownVote(vote)) {
   continue;
  }

  if (!verifyRewardVote(vote)) {
      // handle error
  }

  rewards_votes.AddVote(vote)
}

Reply:
Then we need separate task to do it as this is just chore to fix specific case not re-work rewards votes

Reply:
btw shouldnt it be separate class ? It could nice simple class instead of having it inside votemanager....